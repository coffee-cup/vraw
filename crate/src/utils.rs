#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Pos {
    pub line: u32,
    pub column: u32,
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Range {
    pub start: Pos,
    pub end: Pos,
}

pub trait HasPos {
    fn pos(&self) -> Pos;
}

pub fn create_pos(line: u32, column: u32) -> Pos {
    Pos {
        line: line,
        column: column,
    }
}

pub fn create_range(start: Pos, end: Pos) -> Range {
    Range {
        start: start,
        end: end,
    }
}

pub fn set_panic_hook() {
    // When the `console_error_panic_hook` feature is enabled, we can call the
    // `set_panic_hook` function at least once during initialization, and then
    // we will get better error messages if our code ever panics.
    //
    // For more details see
    // https://github.com/rustwasm/console_error_panic_hook#readme
    #[cfg(feature = "console_error_panic_hook")]
    console_error_panic_hook::set_once();
}
