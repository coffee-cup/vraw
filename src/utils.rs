#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Pos {
    pub line: u32,
    pub column: u32,
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Range {
    pub start: Pos,
    pub end: Pos,
}
