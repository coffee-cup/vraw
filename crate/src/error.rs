use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error<T: fmt::Display> {
    pub error_type: T,
    pub pos: Pos,
}

impl<T: fmt::Display> Error<T> {
    pub fn new(error_type: T, pos: Pos) -> Error<T> {
        Error {
            error_type: error_type,
            pos: pos,
        }
    }
}

impl<T: fmt::Display> HasPos for Error<T> {
    fn pos(&self) -> Pos {
        self.pos
    }
}

impl<T: fmt::Display> fmt::Display for Error<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "[{}:{}]: {}",
            self.pos.line, self.pos.column, self.error_type
        )
    }
}
