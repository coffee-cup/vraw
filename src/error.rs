use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error {
    message: String,
    pos: Option<Pos>,
}

impl Error {
    pub fn new(message: String, pos: Option<Pos>) -> Error {
        Error {
            message: message,
            pos: pos,
        }
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.pos {
            Some(p) => write!(f, "[{}:{}]: {}", p.line, p.column, self.message),
            _ => write!(f, "{}", self.message),
        }
    }
}
