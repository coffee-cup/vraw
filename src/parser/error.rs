use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum ParseErrorType {
    UnExpectedEndOfInput,
    IdentiferCannotBeReservedWord(String),
    UnBalancedParen,
    Expected(String, Option<String>),
}

#[derive(Debug, PartialEq)]
pub struct ParseError {
    pub error_type: ParseErrorType,
    pub pos: Pos,
}

impl HasPos for ParseError {
    fn pos(&self) -> Pos {
        self.pos
    }
}

impl fmt::Display for ParseErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            ParseErrorType::UnExpectedEndOfInput => write!(f, "Unexpected end of input."),
            ParseErrorType::IdentiferCannotBeReservedWord(id) => {
                write!(f, "Identifier {} cannot be a reserved word.", id)
            }
            ParseErrorType::UnBalancedParen => write!(f, "Unbalanced paren."),
            ParseErrorType::Expected(expected, found) => match found {
                Some(found) => write!(f, "Expected {}. Found {}.", expected, found),
                None => write!(f, "Expected {}", expected),
            },
        }
    }
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "[{}:{}] {}",
            self.pos.line, self.pos.column, self.error_type
        )
    }
}
