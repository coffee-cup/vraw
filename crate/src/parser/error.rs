use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum ParseErrorType {
    UnExpectedEndOfInput,
    IdentiferCannotBeReservedWord(String),
    UnBalancedParen,
    Expected(String, Option<String>),
}

impl fmt::Display for ParseErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            ParseErrorType::UnExpectedEndOfInput => write!(f, "Unexpected end of input."),
            ParseErrorType::IdentiferCannotBeReservedWord(id) => {
                write!(f, "Identifier {} cannot be a reserved word.", id)
            }
            ParseErrorType::UnBalancedParen => write!(f, "Unbalanced paren."),
            ParseErrorType::Expected(expected, found) => match found {
                Some(found) => write!(f, "Expected {}. Found {}.", expected, found),
                None => write!(f, "Expected {}", expected),
            },
        }
    }
}

pub type ParseError = Error<ParseErrorType>;
