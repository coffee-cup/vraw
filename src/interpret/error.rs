use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq, Clone)]
pub enum EvalErrorType {
    VariableNotDefined(String),
    ShapeNotDefined(String),
    TypeMismatch(String, String),
    SvgExpectsString(String),
    ShapeAlreadyDefined(String),
    NumArgs(String, usize, usize),
    InvalidArgName(String, String),
    MissingArgs(String, Vec<String>),
    MissingMain,
}

impl fmt::Display for EvalErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            EvalErrorType::VariableNotDefined(name) => write!(f, "Variable `{}` not defined", name),
            EvalErrorType::ShapeNotDefined(name) => write!(f, "Shape `{}` not defined", name),
            EvalErrorType::TypeMismatch(expected, received) => write!(
                f,
                "Typemismatch. Expected: {}, Received: {}",
                expected, received
            ),
            EvalErrorType::SvgExpectsString(received) => write!(
                f,
                "svg value arg needs to be a string. Received {}",
                received
            ),
            EvalErrorType::ShapeAlreadyDefined(name) => write!(f, "Shape {} already defined", name),
            EvalErrorType::NumArgs(func, expected, received) => write!(
                f,
                "Incorrect number of args to {}. Expected: {}, Received: {}",
                func, expected, received
            ),
            EvalErrorType::InvalidArgName(func, arg) => {
                write!(f, "{} does not have an arg named {}", func, arg)
            }
            EvalErrorType::MissingArgs(func, args) => {
                write!(f, "Missing args {} for {}", args.join(", "), func)
            }
            EvalErrorType::MissingMain => write!(f, "Missing main shape"),
        }
    }
}

pub type EvalError = Error<EvalErrorType>;
