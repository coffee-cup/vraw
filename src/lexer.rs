use std::iter::Peekable;

use crate::error::*;
use crate::utils::*;

#[derive(Debug, PartialEq, Clone)]
pub enum TokenType {
    LParen,
    RParen,
    LCurly,
    RCurly,
    Times,
    Divide,
    Plus,
    Minus,
    Equals,
    Compare,
    Colon,
    Comma,
    Number(f64),
    Ident(String),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Token {
    token_type: TokenType,
    token_pos: Range,
}

pub struct Lexer<'a> {
    iter: Peekable<std::str::Chars<'a>>,
    line: u32,
    column: u32,
}

pub fn lex(input: &String) -> Result<Vec<Token>, Error> {
    let mut tokens = vec![];
    let mut lexer = Lexer::new(input);

    while let Some(t) = lexer.next()? {
        tokens.push(t);
    }

    Ok(tokens)
}

impl Token {
    pub fn token_type(&self) -> TokenType {
        self.token_type.clone()
    }

    pub fn token_pos(&self) -> Range {
        self.token_pos
    }
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a String) -> Lexer<'a> {
        Lexer {
            iter: input.trim().chars().peekable(),
            line: 0,
            column: 0,
        }
    }

    fn pos(&self) -> Pos {
        Pos {
            line: self.line,
            column: self.column,
        }
    }

    fn token(&self, token_type: TokenType, start: Pos, end: Pos) -> Option<Token> {
        Some(Token {
            token_type: token_type,
            token_pos: Range {
                start: start,
                end: end,
            },
        })
    }

    fn advance(&mut self, token_type: TokenType) -> Result<Option<Token>, Error> {
        let start = self.pos();

        self.iter.next();
        self.column += 1;

        let end = self.pos();

        let token = self.token(token_type, start, end);
        Ok(token)
    }

    fn advance_whitespace(&mut self) -> bool {
        match self.iter.peek() {
            Some(' ') | Some('\t') => {
                self.column += 1;
                self.iter.next();
                true
            }
            Some('\n') | Some('\r') => {
                self.column = 0;
                self.line += 1;
                self.iter.next();
                true
            }
            _ => false,
        }
    }

    fn forward(&mut self) -> Option<char> {
        self.column += 1;
        self.iter.next()
    }

    fn match_next(&mut self, pred: &Fn(char) -> bool) -> Option<char> {
        let c = match self.iter.peek() {
            Some(c) => *c,
            None => return None,
        };

        if pred(c) {
            self.forward();
            Some(c)
        } else {
            None
        }
    }

    fn consume_ident(&mut self) -> Result<Option<Token>, Error> {
        let start = self.pos();
        let mut id = String::new();

        match self.match_next(&is_alpha) {
            Some(c) => id.push(c),
            None => {
                return Err(Error::new(
                    "identifiers must start with letter.".to_owned(),
                    Some(self.pos()),
                ))
            }
        };

        while let Some(c) = self.match_next(&is_alphanum) {
            id.push(c);
        }
        let end = self.pos();

        let token = self.token(TokenType::Ident(id), start, end);
        Ok(token)
    }

    fn consume_number(&mut self) -> Result<Option<Token>, Error> {
        let start = self.pos();

        let mut str = String::new();
        let mut seen_dot = false;

        while let Some(c) = self.iter.peek() {
            if *c == '.' && !seen_dot {
                seen_dot = true;
                str.push(self.forward().unwrap());
            } else if is_digit(*c) {
                str.push(self.forward().unwrap());
            } else {
                break;
            }
        }

        let end = self.pos();

        let n = str.parse::<f64>().unwrap();

        Ok(self.token(TokenType::Number(n), start, end))
    }

    fn next(&mut self) -> Result<Option<Token>, Error> {
        // ignore whitespace
        while self.advance_whitespace() {}

        let c = match self.iter.peek() {
            Some(c) => c,
            None => return Ok(None),
        };

        let token = match c {
            '(' => self.advance(TokenType::LParen),
            ')' => self.advance(TokenType::RParen),
            '{' => self.advance(TokenType::LCurly),
            '}' => self.advance(TokenType::RCurly),
            '*' => self.advance(TokenType::Times),
            '/' => self.advance(TokenType::Divide),
            '+' => self.advance(TokenType::Plus),
            '-' => self.advance(TokenType::Minus),
            ':' => self.advance(TokenType::Colon),
            ',' => self.advance(TokenType::Comma),
            '=' => {
                let start = self.pos();
                self.forward();
                let end = self.pos();
                match self.iter.peek() {
                    Some('=') => self.advance(TokenType::Compare),
                    _ => Ok(self.token(TokenType::Equals, start, end)),
                }
            }
            'a'...'z' => self.consume_ident(),
            '0'...'9' => self.consume_number(),
            _ => Err(Error::new(
                format!("unexpected character: {:?}", c),
                Some(self.pos()),
            )),
        };

        token
    }
}

fn is_alpha(c: char) -> bool {
    match c {
        'a'...'z' | 'A'...'Z' => true,
        _ => false,
    }
}

fn is_digit(c: char) -> bool {
    match c {
        '0'...'9' => true,
        _ => false,
    }
}

fn is_alphanum(c: char) -> bool {
    is_alpha(c) || is_digit(c)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lex_identifier() {
        let tokens = lex(&"a1234".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Ident("a1234".to_owned()),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 5 }
                }
            }])
        );
    }

    #[test]
    fn lex_integer() {
        let tokens = lex(&"10".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Number(10.0),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 2 }
                }
            }])
        )
    }

    #[test]
    fn lex_float() {
        let tokens = lex(&"10.123".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Number(10.123),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 6 }
                }
            }])
        )
    }
}
