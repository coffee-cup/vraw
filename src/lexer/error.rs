use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum LexerErrorType {
    InvalidIdentifier,
    StringNeverTerminated,
    UnexpectedCharacter(char),
}

impl fmt::Display for LexerErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            LexerErrorType::InvalidIdentifier => write!(f, "Identifier must start with a letter"),
            LexerErrorType::StringNeverTerminated => write!(f, "String is never terminated"),
            LexerErrorType::UnexpectedCharacter(c) => write!(f, "Unexpected character `{}`", c),
        }
    }
}

pub type LexerError = Error<LexerErrorType>;
