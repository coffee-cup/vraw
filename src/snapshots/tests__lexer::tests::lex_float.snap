---
created: "2019-06-25T19:31:56.731728Z"
creator: insta@0.8.1
source: src/lexer.rs
expression: tokens
---
Ok(
    [
        Token {
            token_type: Number(
                10.123
            ),
            token_pos: Range {
                start: Pos {
                    line: 0,
                    column: 0
                },
                end: Pos {
                    line: 0,
                    column: 6
                }
            }
        }
    ]
)
